### PR TITLE
Use optimized queries

### LINKED ISSUE #205: Possible EventEmitter memory leak detected
**Description**
After some time of my browser being in the background I went to use Multix and noticed a generic error being displayed on the main screen. 

After viewing the console it implies there is a "Possible EventEmitter memory leak detected" 

**Steps To Reproduce**

Use multix and then leave it backgrounded, return to it an hour or so later (not 100% sure about these steps yet). 

**Expected Behaviour**
Multix should display as normal 

**Console Error**

Screenshot 2023-06-13 at 12 49 50 PM

Full console dump:
[console_log.txt](https://github.com/ChainSafe/Multix/files/11738847/console_log.txt)

**Screenshots / Video**
Screenshot 2023-06-13 at 12 50 54 PM

**Test Host Specifics:**

- OS: macOS
- Firefox 

Not sure if this issue is OS or browser dependent but I've only seen it happen with this combo at the moment. 

**Additional context**
- As soon as I refreshed the browser there was no longer an issue so this is not hugely user-impacting

### LINKED ISSUE #268: Optimize squid queries
**Description**
Graphql query to fetch all multisigs is quite complex and generates very unoptimized SQL query. To make things worse, when we do subscription, it continuously executes that unoptimized SQL query every 5 seconds. We haven't noticed that before because every network was in it's own database schema which meant there was a lot less rows in tables.
Current grapql query: https://github.com/ChainSafe/Multix/blob/86c486945c2ac4223ab682c68bc0204fd09f43fc/packages/ui/src/queries/multisigsByAccounts.graphql

**Prerequisites**

Regular usage

**Steps To Reproduce**

Open multix

**Expected Behaviour**

Super fast response and optimized query

**Potential Solutions**

### 1. Separate everything into different databases.
 - Pro:
    - "simple" solution that will buy us some time
 - Con:
    - once aquarium introduces billing it will be expensive for us
    - we need to change frontend to query different endpoints for each network

### 2. Optimize query

The following query generates two very optimized and fast sql queries which allows us to get multisig accounts. For delagator/delagatee stuff we would need to jump on a call with whiteboard to challge what we actually need.

```graphql
query MultisigsByAccounts($accountIds: [String!], $watchedAccountIds: [String!]) {
  accountMultisigs(where: { signatory: { id_in: $accountIds } }) {
    multisig {
      address,
      isMultisig,
      isPureProxy,
      signatories {
        id
      }

    }
  }
  accounts(where: {id_in: $watchedAccountIds     AND: { isMultisig_eq: true, OR: { isPureProxy_eq: true } }}) {
      address,
      isMultisig,
      isPureProxy,
      signatories {
        id
      }
  }
}
```

Note: for subscription, we would need to send that as two separate subscription querys.

### 3. Keep same db but generate models for each network

For example on squid side, instead of:
```graphql
type Account @entity {
  id: ID! @index @unique
  address: String!
  multisigs: [AccountMultisig] @derivedFrom(field: "signatory")
  isPureProxy: Boolean @index
  # it's the origin/delegator for another proxy accounts, a pure must have some
  delegatorFor: [ProxyAccount!] @derivedFrom(field: "delegator")
  # it's the delegatee/doing stuff on behalf of another account
  delegateeFor: [ProxyAccount!] @derivedFrom(field: "delegatee")
  isMultisig: Boolean @index
  signatories: [AccountMultisig!] @derivedFrom(field: "multisig")
  threshold: Int
  multisigsCalls: [MultisigCall!] @derivedFrom(field: "multisig")
}

```

we would have:
```graphql
type PolkadotAccount @entity {
  id: ID! @index @unique
  address: String!
  multisigs: [PolkadotAccountMultisig] @derivedFrom(field: "signatory")
  isPureProxy: Boolean @index
  # it's the origin/delegator for another proxy accounts, a pure must have some
  delegatorFor: [PolkadotProxyAccount!] @derivedFrom(field: "delegator")
  # it's the delegatee/doing stuff on behalf of another account
  delegateeFor: [PolkadotProxyAccount!] @derivedFrom(field: "delegatee")
  isMultisig: Boolean @index
  signatories: [PolkadotAccountMultisig!] @derivedFrom(field: "multisig")
  threshold: Int
  multisigsCalls: [PolkadotMultisigCall!] @derivedFrom(field: "multisig")
}

type RococoAccount @entity {
  id: ID! @index @unique
  address: String!
  multisigs: [RococoAccountMultisig] @derivedFrom(field: "signatory")
  isPureProxy: Boolean @index
  # it's the origin/delegator for another proxy accounts, a pure must have some
  delegatorFor: [RococoProxyAccount!] @derivedFrom(field: "delegator")
  # it's the delegatee/doing stuff on behalf of another account
  delegateeFor: [RococoProxyAccount!] @derivedFrom(field: "delegatee")
  isMultisig: Boolean @index
  signatories: [RococoAccountMultisig!] @derivedFrom(field: "multisig")
  threshold: Int
  multisigsCalls: [RococoMultisigCall!] @derivedFrom(field: "multisig")
}
//etc

```
and then in processor we would dynamically load models based on network.
For example:
```typescript
import * as DbModels from "../model"

const Account = DbModels[chainId+"Account"];

await ctx.store.save(new Account({...}));
```

### PR DESCRIPTION
**Pull request by Tbaut**
- closes #268 
- closes #205 by implementing a `refetch` on error. 
- Set the pure to appear first in the list.

This implements the 2nd suggestion to have optimized queries to subsquid.
It adds a nice to have that if you add a watch account that's a signatory for 1 multisig, with 1 proxy above, you'll end up actually seeing in the Multix all the other multsigs that the proxy has below it. It wasn't the case before.

eg, on Kusama `Alice u Bob` with address `DGaHm71gRHJGQVF2ubunkx37qgLpno2gYibq2WHoDJpFMt6` is only part of 1 multisig, that has the ChaosDAO pure above `DCZyhphXsRLcW84G9WmWEXtAA8DKGtVGSFZLJYty8Ajjyfa`. If you add this former account as watched, you'll see the ChaosDAO pure along with all the other multisigs below it as if you were watching the pure.

Regarding testing, this is reworking the whole multisig queries and logic so the best way to test is:
- using watch account, know what to expect and add an account. Verify that the proxy/multisigs that you see are the ones expected.
- using your own accounts, verify that the dropdown displays the same multisigs/pure as on prod.
- I launched a huge subsquid like the one that was very slow and this branch is deployed against it. we should try to see if things are slow at times.

### CODE REVIEW

**Comment by Tbaut** [comment]
I launched a huge squid like the one that was very slow the past week and this branch is deployed against it. we should try to see if things are slow at times.

**Comment by asnaith** [comment]
🟢 I can now see that all the other multisigs under the pure, not just the multisig the watched signatory is part of
🟢 When switching between networks all resolved within 2 - 5 seconds (saw the same on prod too)
🟢 Verified that my connected accounts/dropdown looked the same as prod
🟢 I see the pure as the first option in the dropdown list now. However....
🟠 When we add a signatory to watch, the multisig is still the default selection in the dropdown not the pure. Is this expected?

**Comment by Tbaut** [comment]
About your last point, it's indeed expected. We have no logic to automatically switch to a given account because
- how should we chose one if there are many, and how should the user know that there are many
- this logic would definitely be a little hacky, since the whole UI is reactive and just adapts to new changes automatically.

The best way I can think of to handle this would be something like
- make a dedicated 1 time query to see if there are pure/msig linked to this account
- inform the user (in a modal?) that there are now x new accounts to explore in the drop-down, we could offer a way to switch to them with 1 click.

**Comment by Tbaut** [comment]
FYI @asnaith I saw this time my fix working for the re-connection to the indexer after a while. The error message to look for is "Subscription [some id] error, refetching"

File: packages/ui/src/contexts/MultiProxyContext.tsx [hunk 1]
```diff
@@ -64,136 +76,116 @@ const MultiProxyContextProvider = ({ children }: MultisigContextProps) => {
...
+                  signatories: getSignatoriesFromAccount(multisig.signatories),
+                  threshold: multisig.threshold
+                }
+              ]
+            } as MultiProxy)
+          }
+        })
+
+        // flatten out multisigMap
+        const multisigArray = Array.from(multisigMap.entries()).map(([, multip]) => multip)
```
Comment (mpetrunic) [inline packages/ui/src/contexts/MultiProxyContext.tsx:117]:
```suggestion
        const multisigArray = Array.from(multisigMap.values())
```
wouldn't that do same thing without allocating array twice? :)

File: packages/ui/src/contexts/MultiProxyContext.tsx [hunk 2]
```diff
@@ -64,136 +76,116 @@ const MultiProxyContextProvider = ({ children }: MultisigContextProps) => {
...
+        // find the delegatee that are multisigs and put all the infos right away
+        account.delegatorFor.forEach(({ delegatee, type }) => {
+          if (delegatee.isMultisig) {
+            const previousMultisigsForProxy = pureProxyMap.get(account.address)?.multisigs || []
 
-        // if this account is a multisig and is the delegatee for at least a pureProxy
-        if (account.isMultisig && pureProxyAddresses?.length > 0) {
-          pureProxyAddresses.forEach(({ pureAddress, type }) => {
-            const previousMultisigsForProxy = pureProxyMap.get(pureAddress)?.multisigs || []
+            const isAlreadyInMultisigList = !!previousMultisigsForProxy.find(
```
Comment (mpetrunic) [inline packages/ui/src/contexts/MultiProxyContext.tsx:142]:
```suggestion
            const isAlreadyInMultisigList = previousMultisigsForProxy.some(
```
More optimized as it returns immediately first time condition is true^^

File: packages/ui/src/contexts/MultiProxyContext.tsx [hunk 3]
```diff
@@ -64,136 +76,116 @@ const MultiProxyContextProvider = ({ children }: MultisigContextProps) => {
...
-        ([proxy, agg]) =>
-          ({
-            proxy,
-            multisigs: agg.multisigs
-          } as MultiProxy)
-      )
-
-      res.push(...proxyArray)
+    // flatten out pureProxyMap
+    const pureProxyArray = Array.from(pureProxyMap.entries()).map(
```
Comment (mpetrunic) [inline packages/ui/src/contexts/MultiProxyContext.tsx:165]: again .values() will save on memory allocation

Comment (Tbaut) [reply packages/ui/src/contexts/MultiProxyContext.tsx:165]: absolutely

File: packages/ui/src/contexts/MultiProxyContext.tsx (hunk 1 above)
Comment (Tbaut) [reply packages/ui/src/contexts/MultiProxyContext.tsx:117]: nice catch, yes

File: packages/ui/src/contexts/MultiProxyContext.tsx (hunk 3 above)
Comment (Tbaut) [reply packages/ui/src/contexts/MultiProxyContext.tsx:165]: ah not for that one, we need both key and value here.

Comment (mpetrunic) [reply packages/ui/src/contexts/MultiProxyContext.tsx:165]: ah true, but why do you need Array.from then?

Comment (Tbaut) [reply packages/ui/src/contexts/MultiProxyContext.tsx:165]: the `.entries()` doesn't return an array, it returns an `IterableIterator<[string, Omit<MultiProxy, "proxy">]>` so you have to make this an array to be able to call `.map` on it.

**Review by mpetrunic — APPROVED** [review]

**Review by asnaith — APPROVED** [review]

**Comment by Tbaut** [comment]
I'll keep the old indexers running for the next days, so that we can revert quickly if needed.

**Pull request merged**